### PR TITLE
Solve eagerness issues for `MERGE`

### PR DESCRIPTION
**Pull request by Mats-SX**
This PR introduces slight regressions in the amount of eagerness we plan versus the optimum.

### CODE REVIEW

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanSingleQuery.scala [hunk 1]
```diff
@@ -58,67 +59,67 @@ case class PlanSingleQuery(planPart: (PlannerQuery, LogicalPlanningContext, Opti
...
-   * - (a)-[r]-(b) (undirected)
-   * - (a)-[r1]->(b)-[r2]->(c) (multi step)
-   * - (a)-[r*]->(b) (variable length)
-   */
-  private def hasUnsafeRelationships(pq: PlannerQuery) = {
-    val allPatterns = pq.queryGraph.allPatternRelationships
-    allPatterns.size > 1 ||
-      allPatterns.exists(r => r.dir == SemanticDirection.BOTH || !r.length.isSimple)
-  }
+//  private def conflicts(plan: LogicalPlan, plannerQuery: PlannerQuery): Boolean = {
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanSingleQuery.scala:62]: remove

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanWithTail.scala [hunk 2]
```diff
@@ -66,32 +64,51 @@ case class PlanWithTail(expressionRewriterFactory: (LogicalPlanningContext => Re
...
-        val partPlan = planPart(query, lhsContext, Some(context.logicalPlanProducer.planQueryArgumentRow(query.queryGraph)))
+        val partPlan = planPart(plannerQuery, lhsContext, Some(context.logicalPlanProducer.planQueryArgumentRow(plannerQuery.queryGraph)))
         ///use eager if configured to do so
         val alwaysEager = context.config.updateStrategy.alwaysEager
         //If reads interfere with writes, make it a RepeatableRead
+        var shouldPlanEagerBeforeTail = false
         val planWithEffects =
-          if (alwaysEager || (query.updateGraph overlaps query.queryGraph))
-            context.logicalPlanProducer.planRepeatableRead(partPlan)
+          if (!plannerQuery.isInstanceOf[MergePlannerQuery] &&
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanWithTail.scala:75]: This part is pretty hard to grasp, could we simplify? Getting rid of `isInstanceOf` would be nice as well.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanWithTail.scala [hunk 3]
```diff
@@ -66,32 +64,51 @@ case class PlanWithTail(expressionRewriterFactory: (LogicalPlanningContext => Re
...
 
         //If previous update interferes with any of the reads here or in tail, make it an EagerApply
         val applyPlan = {
           val lastPlannerQuery = lhs.solved.last
-          val newLhs = if (alwaysEager ||
-            (!lastPlannerQuery.writeOnly && query.allQueryGraphs.exists(lastPlannerQuery.updateGraph.overlaps)))
+          val newLhs =
+            if (alwaysEager ||
+              plannerQuery.allUpdateGraphs.exists(lastPlannerQuery.updateGraph.deleteOverlapWithMergeIn) ||
+              (!lastPlannerQuery.writeOnly && /*!plannerQuery.isInstanceOf[MergePlannerQuery] &&*/
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanWithTail.scala]: remove commented out code

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanWithTail.scala (hunk 2 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanWithTail.scala:75]: Agreed. This will be fixed in a separate PR.
